From 7c85e2e363fd39e60bf5041d02e14e8bd62c1a68 Mon Sep 17 00:00:00 2001
From: Quirin Gylstorff <quirin.gylstorff@siemens.com>
Date: Tue, 24 Mar 2020 17:58:08 +0100
Subject: [PATCH v7 1/3] meta/support: Generate a custom initramfs

This package sets the Parameters for mkinitramfs/update-intramfs
before it regenerates the initrd.img of debian with a modified version.

Use cases are the remove unnecessary kernel modules to reduce the
size of the initrd by using the parameters:
```
INITRAMFS_MODULES = "list"
INITRAMFS_MODULE_LIST += "ext4"
```

Set the boot root during the initrd generation by setting `INITRAMFS_ROOT`.

see also man pages of mkinitramfs and initramfs.conf.

Signed-off-by: Quirin Gylstorff <quirin.gylstorff@siemens.com>
---
 .../initramfs-config/initramfs-config_0.1.bb  |  6 +++
 .../initramfs-config/files/control.tmpl       | 12 +++++
 .../initramfs-config/files/postinst.tmpl      | 50 +++++++++++++++++++
 .../initramfs-config/files/postrm.tmpl        | 41 +++++++++++++++
 .../initramfs-config/initramfs-config.inc     | 32 ++++++++++++
 5 files changed, 141 insertions(+)
 create mode 100644 meta-isar/recipes-support/initramfs-config/initramfs-config_0.1.bb
 create mode 100644 meta/recipes-support/initramfs-config/files/control.tmpl
 create mode 100644 meta/recipes-support/initramfs-config/files/postinst.tmpl
 create mode 100644 meta/recipes-support/initramfs-config/files/postrm.tmpl
 create mode 100644 meta/recipes-support/initramfs-config/initramfs-config.inc

diff --git a/meta-isar/recipes-support/initramfs-config/initramfs-config_0.1.bb b/meta-isar/recipes-support/initramfs-config/initramfs-config_0.1.bb
new file mode 100644
index 0000000..c951e8a
--- /dev/null
+++ b/meta-isar/recipes-support/initramfs-config/initramfs-config_0.1.bb
@@ -0,0 +1,6 @@
+#
+# Copyright (C) Siemens AG, 2020
+#
+# SPDX-License-Identifier: MIT
+
+require recipes-support/initramfs-config/initramfs-config.inc
diff --git a/meta/recipes-support/initramfs-config/files/control.tmpl b/meta/recipes-support/initramfs-config/files/control.tmpl
new file mode 100644
index 0000000..66984eb
--- /dev/null
+++ b/meta/recipes-support/initramfs-config/files/control.tmpl
@@ -0,0 +1,12 @@
+Source: ${PN}
+Section: misc
+Priority: optional
+Standards-Version: 3.9.6
+Maintainer: isar-users <isar-users@googlegroups.com>
+Build-Depends: debhelper (>= 9)
+
+
+Package: ${PN}
+Architecture: any
+Depends: ${shlibs:Depends}, ${misc:Depends}, initramfs-tools-core, ${DEBIAN_DEPENDS}
+Description: Configuration files for a custom initramfs
diff --git a/meta/recipes-support/initramfs-config/files/postinst.tmpl b/meta/recipes-support/initramfs-config/files/postinst.tmpl
new file mode 100644
index 0000000..e523906
--- /dev/null
+++ b/meta/recipes-support/initramfs-config/files/postinst.tmpl
@@ -0,0 +1,50 @@
+#!/bin/sh
+# postinst script for initramfs-config
+#
+# see: dh_installdeb(1)
+
+set -e
+
+case "$1" in
+    configure)
+        INITRAMFS_CONF=/etc/initramfs-tools/initramfs.conf
+        if [ -f ${INITRAMFS_CONF} ]; then
+            sed -i -E 's/(^MODULES=).*/\1${INITRAMFS_MODULES}/' ${INITRAMFS_CONF}
+            sed -i -E 's/(^BUSYBOX=).*/\1${INITRAMFS_BUSYBOX}/' ${INITRAMFS_CONF}
+            sed -i -E 's/(^COMPRESS=).*/\1${INITRAMFS_COMPRESS}/' ${INITRAMFS_CONF}
+            sed -i -E 's/(^KEYMAP=).*/\1${INITRAMFS_KEYMAP}/' ${INITRAMFS_CONF}
+            sed -i -E 's/(^DEVICE=).*/\1${INITRAMFS_NET_DEVICE}/' ${INITRAMFS_CONF}
+            sed -i -E 's/(^NFSROOT=).*/\1${INITRAMFS_NFSROOT}/' ${INITRAMFS_CONF}
+            sed -i -E 's/(^RUNSIZE=).*/\1${INITRAMFS_RUNSIZE}/' ${INITRAMFS_CONF}
+            if grep -Fxq "ROOT=" "${INITRAMFS_CONF}"; then
+                sed -i -E 's/(^ROOT=).*/\1${INITRAMFS_ROOT}/' ${INITRAMFS_CONF}
+            else
+                sed -i -E "\$aROOT=${INITRAMFS_ROOT}" ${INITRAMFS_CONF}
+            fi
+        fi
+
+        MODULES_LIST_FILE=/etc/initramfs-tools/modules
+        if [ -f ${MODULES_LIST_FILE} ]; then
+            for modname in ${INITRAMFS_MODULE_LIST}; do
+                if ! grep -Fxq "$modname" "${MODULES_LIST_FILE}"; then
+                    echo "$modname" >> "${MODULES_LIST_FILE}"
+                fi
+            done
+        fi
+
+        update-initramfs -v -u
+
+    ;;
+    abort-upgrade|abort-remove|abort-deconfigure)
+    ;;
+
+    *)
+        echo "postinst called with unknown argument \`$1'" >&2
+        exit 1
+        ;;
+esac
+# dh_installdeb will replace this with shell code automatically
+# generated by other debhelper scripts.
+#DEBHELPER#
+
+exit 0
diff --git a/meta/recipes-support/initramfs-config/files/postrm.tmpl b/meta/recipes-support/initramfs-config/files/postrm.tmpl
new file mode 100644
index 0000000..115d9b6
--- /dev/null
+++ b/meta/recipes-support/initramfs-config/files/postrm.tmpl
@@ -0,0 +1,41 @@
+#!/bin/sh
+# postrm script for initramfs-config
+#
+# see: dh_installdeb(1)
+
+set -e
+
+case "$1" in
+    purge|remove|upgrade|failed-upgrade|abort-install|abort-upgrade|disappear)
+        # back to the debian defaults
+        INITRAMFS_CONF=/etc/initramfs-tools/initramfs.conf
+        sed -i -E 's/(^MODULES=).*/\1most/' ${INITRAMFS_CONF}
+        sed -i -E 's/(^BUSYBOX=).*/\1auto/' ${INITRAMFS_CONF}
+        sed -i -E 's/(^COMPRESS=).*/\1gzip/' ${INITRAMFS_CONF}
+        sed -i -E 's/(^KEYMAP=).*/\1n/' ${INITRAMFS_CONF}
+        sed -i -E 's/(^DEVICE=).*/\1/' ${INITRAMFS_CONF}
+        sed -i -E 's/(^NFSROOT=).*/\1auto/' ${INITRAMFS_CONF}
+        sed -i -E 's/(^RUNSIZE=).*/\110%/' ${INITRAMFS_CONF}
+        sed -i -E 's/(^ROOT=).*//' ${INITRAMFS_CONF}
+
+        # remove the added modules
+        MODULES_LIST_FILE=/etc/initramfs-tools/modules
+        for modname in ${INITRAMFS_MODULE_LIST}; do
+            sed -i -E 's/$modname//'
+        done
+
+        update-initramfs -v -u
+    ;;
+
+    *)
+        echo "postrm called with unknown argument \`$1'" >&2
+        exit 1
+        ;;
+esac
+
+# dh_installdeb will replace this with shell code automatically
+# generated by other debhelper scripts.
+
+#DEBHELPER#
+
+exit 0
diff --git a/meta/recipes-support/initramfs-config/initramfs-config.inc b/meta/recipes-support/initramfs-config/initramfs-config.inc
new file mode 100644
index 0000000..16049a9
--- /dev/null
+++ b/meta/recipes-support/initramfs-config/initramfs-config.inc
@@ -0,0 +1,32 @@
+# This software is a part of ISAR.
+# Copyright (C) 2020 Siemens AG
+#
+# SPDX-License-Identifier: MIT
+inherit dpkg-raw
+inherit template
+DESCRIPTION = "Recipe to set the initramfs configuration and generate a new ramfs"
+
+FILESEXTRAPATHS_prepend := "${FILE_DIRNAME}/files:"
+
+SRC_URI = "file://postinst.tmpl \
+           file://postrm.tmpl \
+           file://control.tmpl \
+          "
+
+INITRAMFS_MODULES ?= "most"
+INITRAMFS_BUSYBOX ?= "auto"
+INITRAMFS_COMPRESS ?= "gzip"
+INITRAMFS_KEYMAP ?= "n"
+INITRAMFS_NET_DEVICE ?= ""
+INITRAMFS_NFSROOT ?= "auto"
+INITRAMFS_RUNSIZE ?= "10%"
+INITRAMFS_ROOT ?= ""
+INITRAMFS_MODULE_LIST ?= ""
+CREATE_NEW_INITRAMFS ?= "n"
+KERNEL_PACKAGE = "${@ ("linux-image-" + d.getVar("KERNEL_NAME", True)) if d.getVar("KERNEL_NAME", True) else ""}"
+DEBIAN_DEPENDS += ", ${KERNEL_PACKAGE}"
+TEMPLATE_FILES = "postinst.tmpl control.tmpl postrm.tmpl"
+TEMPLATE_VARS += "INITRAMFS_MODULES INITRAMFS_BUSYBOX INITRAMFS_COMPRESS \
+                  INITRAMFS_KEYMAP INITRAMFS_NET_DEVICE INITRAMFS_NFSROOT \
+                  INITRAMFS_RUNSIZE INITRAMFS_ROOT INITRAMFS_MODULE_LIST \
+                  CREATE_NEW_INITRAMFS DEBIAN_DEPENDS PN"
-- 
2.20.1

